lib/benchmarks/gallery_automator.dart

```diff
@@ -0,0 +1,199 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+import 'package:gallery/data/demos.dart';
+import 'package:gallery/l10n/gallery_localizations_en.dart';
+import 'package:gallery/main.dart';
+
+import 'scroll.dart';
+
+const Duration _initialWaitingDuration = Duration(milliseconds: 1500);
+
+const List<String> _demosWithAnimation = <String>[
+  'Progress indicators@material',
+  'Activity indicator@cupertino',
+  'Colors@other',
+];
+
+const Duration _defaultWaitingDuration = Duration(seconds: 3);
+
+enum DemoType {
+  study,
+  animatedWidget,
+  unanimatedWidget,
+}
+
+DemoType typeOfDemo(String demo) {
+  if (demo.contains('@study')) {
+    return DemoType.study;
+  } else if (_demosWithAnimation.contains(demo)) {
+    return DemoType.animatedWidget;
+  } else {
+    return DemoType.unanimatedWidget;
+  }
+}
+
+/// A class that automates the gallery.
+class GalleryAutomator {
+  GalleryAutomator({
+    @required this.benchmarkName,
+    @required this.reporter,
+    @required this.reportError,
+    @required this.runCriterion,
+    this.testScrollsOnly = false,
+  });
+
+  final String benchmarkName;
+  final void Function(String) reporter;
+  final Future<void> Function(dynamic error, StackTrace stackTrace) reportError;
+  final bool Function(String) runCriterion;
+  final bool testScrollsOnly;
+
+  bool finished = false;
+  LiveWidgetController controller;
+
+  /// An iterable that generates all demo names.
+  Iterable<String> get demoNames =>
+      allGalleryDemos(GalleryLocalizationsEn()).map((demo) => demo.describe);
+
+  Future<void> _handleError(dynamic error, StackTrace stackTrace) async {
+    print('Error: $error');
+    print('StackTrace: $stackTrace');
+    await reportError(error, stackTrace);
+  }
+
+  Widget createWidget() {
+    Future<void>.delayed(
+      _initialWaitingDuration,
+      testScrollsOnly ? automateScrolls : automateDemoGestures,
+    ).catchError(_handleError);
+    return const GalleryApp();
+  }
+
+  /// Opens and quits demos that are specified by [runCriterion], twice.
+  Future<void> automateDemoGestures() async {
+    await animationStops();
+
+    controller = LiveWidgetController(WidgetsBinding.instance);
+
+    reporter('==== List of demos to be run ====');
+    for (final demo in demoNames) {
+      if (!runCriterion(demo)) continue;
```

Comment:
I'm under the impression that using the continue method is bad practice but not too sure

Reply:
Changed to block statement, which is easier to read.